pkg/loop/internal/pb/chain_reader.proto

```diff
@@ -34,3 +77,142 @@ message BoundContract {
   string name = 2;
   bool pending = 3;
 }
+
+enum ComparisonOperator {
+  Eq = 0;
+  Neq = 1;
+  Gt = 2;
+  Lt = 3;
+  Gte = 4;
+  Lte = 5;
+}
+
+message ValueComparator {
+  // TODO pass interface{} here
+  string value = 1;
+  ComparisonOperator operator = 2;
+}
+
+message KeyValuesComparator{
+  string key = 1;
+  repeated ValueComparator value_comparators = 2;
+}
+
+// QueryFilter represents a lightweight orm like DSL defined for filtering over common blockchain primitives.
+message QueryFilter {
+  repeated Expression expression = 1;
+}
+
+message KeyFilter {
+  string key = 1;
+  QueryFilter query_filter = 2;
+}
+
+// Expression encapsulates a single unit of filtering logic, which can be a common blockchain primitive or a composite of boolean expressions. 
+// This allows for both simple and more complex nested expressions.
+message Expression {
+  oneof evaluator {
+    Primitive primitive = 1;
+    BooleanExpression boolean_expression = 2;
+  }
+}
+
+enum BooleanOperator {
+  AND = 0;
+  OR = 1;
+}
+
+message BooleanExpression {
+  BooleanOperator boolean_operator = 1;
+  repeated Expression expression = 2;
+}
+
+message And {
+  repeated Expression a = 1;
+}
+
+message Or {
+  repeated Expression a = 1;
+}
+
+message Address {
+  repeated string addresses = 1;
+}
+
+enum ConfirmationLevel {
+  Finalized   = 0;
+  Unconfirmed = 1;
+}
+
+message Confirmations {
+  ConfirmationLevel confirmations = 1;
+}
+
+message Block{
+  uint64 block_number = 1;
+  ComparisonOperator operator = 2;
+}
+
+message Timestamp{
+  uint64 timestamp = 1;
+  ComparisonOperator operator = 2;
+}
+
+message TxHash{
+  string tx_hash = 1;
+}
+
+// Primitive defines the basic building blocks for filter conditions based around fundamental blockchain concepts.
+message Primitive {
+  oneof comparator {
+    Address address = 3;
+    Confirmations confirmations = 4;
+    Block block = 5;
+    Timestamp timestamp = 6;
+    TxHash tx_hash = 7;
+  }
+}
```

Comment:
I was trying to understand a different comment, where we were discussing `KeyLIst` and `KeyFilter` and it led me here.

I was picturing when we DSL-ified the original `QueryEvents` proposal in the CLIP with KeysList was that there would be another primitive in this list for key filtering, along with `Address`, `Confirmations`, `Block`, `Timestamp`, and `TxHash`.  I was mostly thinking it would just be `Key` but it could also be a`KeyList` if we wanted to handle automatically AND'ing together a list of keys in the same primitive.

Before I mentioned that you split off Key[0] as having special significance from the other keys, but another thing I'm noticing now is that all of the keys are treated different from the other fields you can filter on. Not sure if that's better or worse, but wanted to remind you in case you hadn't thought about it this way, that they don't have to be treated so differently. I'm not sure what the disadvantages may be, but it would make more sense to me to just be able to use keys in the DSL in the same way you use any other of these fields.